compile.c

```diff
@@ -9736,6 +9740,127 @@ compile_attrasgn(rb_iseq_t *iseq, LINK_ANCHOR *const ret, const NODE *const node
     return COMPILE_OK;
 }
 
+static VALUE
+compile_negative_numeric(VALUE val)
+{
+    if (FIXNUM_P(val)) {
+        return LONG2FIX(-FIX2LONG(val));
+    }
+    if (SPECIAL_CONST_P(val)) {
+#if USE_FLONUM
+        if (FLONUM_P(val)) {
+            return DBL2NUM(-RFLOAT_VALUE(val));
+        }
+#endif
+    }
+    switch (OBJ_BUILTIN_TYPE(val))
+    {
+    case T_BIGNUM:
+        BIGNUM_NEGATE(val);
+        val = rb_big_norm(val);
+        break;
+    case T_RATIONAL:
+        RATIONAL_SET_NUM(val, compile_negative_numeric(RRATIONAL(val)->num));
+        break;
+    case T_COMPLEX:
+        RCOMPLEX_SET_REAL(val, compile_negative_numeric(RCOMPLEX(val)->real));
+        RCOMPLEX_SET_IMAG(val, compile_negative_numeric(RCOMPLEX(val)->imag));
+        break;
+    case T_FLOAT:
+        val = DBL2NUM(-RFLOAT_VALUE(val));
+        break;
+    default:
+        val = LONG2FIX(-FIX2LONG(val));
+        break;
+    }
+    return val;
+}
+
+VALUE
+rb_compile_integer_literal(rb_node_integer_t* node)
+{
+    VALUE val = rb_cstr_to_inum(node->val, node->base, FALSE);
+    if (node->tminus) {
+        val = compile_negative_numeric(val);
+    }
+    return val;
+}
+
+VALUE
+rb_compile_float_literal(rb_node_float_t* node)
+{
+    double d = strtod(node->val, 0);
+    VALUE val = DBL2NUM(d);
+    if (node->tminus) {
+        val = compile_negative_numeric(val);
+    }
```

Comment:
Why not simply negate the primitive type?
`compile_negative_numeric` seems unnecessary.

```suggestion
    if (node->tminus) {
        d = -d;
    }
    VALUE val = DBL2NUM(d);
```